### PR TITLE
feat(core/ui): add close button to modals

### PR DESCRIPTION
**Pull request by SuyashSalampuria**

### CODE REVIEW

File: assets/ui.css [hunk 1]
```diff
@@ -1,3 +1,14 @@
+#close-button{
```
Comment (saschanaz) [inline assets/ui.css]:
```suggestion
.close-button {
```

File: assets/ui.css [hunk 2]
```diff
@@ -1,3 +1,14 @@
...
+  font-weight: bold;
+  float: left;
+  cursor: pointer;
+  margin-left: 0.5%;
+}
+
+#close-button:hover {
```
Comment (saschanaz) [inline assets/ui.css]:
```suggestion
.close-button:hover {
```

File: src/core/ui.js [hunk 3]
```diff
@@ -47,13 +47,15 @@ function ariaDecorate(elem, ariaMap) {
 
 const respecUI = hyperHTML`<div id='respec-ui' class='removeOnSave' hidden></div>`;
 const menu = hyperHTML`<ul id=respec-menu role=menu aria-labelledby='respec-pill' hidden></ul>`;
+const closeButton = hyperHTML`<span id="close-button">❌</span>`;
```
Comment (saschanaz) [inline src/core/ui.js]:
```suggestion
const closeButton = hyperHTML`<span class="close-button">❌</span>`;
```

`class` is preferred since we don't use the id anywhere else. Also I would prefer putting this line at line 51 as currently line 49 and 50 are related and should be together.

File: src/core/ui.js [hunk 4]
```diff
@@ -47,13 +47,15 @@ function ariaDecorate(elem, ariaMap) {
...
 window.addEventListener("load", () => trapFocus(menu));
 let modal;
 let overlay;
 const errors = [];
 const warnings = [];
 const buttons = {};
 
+closeButton.addEventListener("click", () => ui.closeModal());
```
Comment (saschanaz) [inline src/core/ui.js]: Maybe move to line 52? I would prefer putting the relevant lines together.

File: src/core/ui.js (hunk 3 above)
Comment (saschanaz) [reply src/core/ui.js]: Can we use `button` here rather than `span`, so that screen readers can see whether this is a button without `aria-role`?

File: assets/ui.css [hunk 5]
```diff
@@ -1,3 +1,14 @@
+.close-button{
```
Comment (saschanaz) [inline assets/ui.css]:
```suggestion
.close-button {
```

File: assets/ui.css [hunk 6]
```diff
@@ -1,3 +1,14 @@
...
+  font-weight: bold;
+  float: left;
+  cursor: pointer;
+  margin-left: 0.5%;
```
Comment (saschanaz) [inline assets/ui.css]:
```suggestion
  margin-left: 0.5%;
  border: none;
  background: transparent;
```

File: assets/ui.css [hunk 7]
```diff
@@ -1,3 +1,14 @@
+.close-button{
+  font-weight: bold;
+  float: left;
+  cursor: pointer;
```
Comment (saschanaz) [inline assets/ui.css]: Not sure this one is needed, @sidvishnoi what do you think?

Comment (SuyashSalampuria) [reply assets/ui.css]: float:left would be fine. I think.

Comment (saschanaz) [reply assets/ui.css]: Yeah, I mean `cursor: pointer` 👀

Comment (SuyashSalampuria) [reply assets/ui.css]: Also, is this margin for the close button fine or i shall decrease it
Screenshot from 2020-01-09 13-48-19

File: assets/ui.css (hunk 6 above)
Comment (saschanaz) [inline assets/ui.css]: Maybe a fixed value, as there is no need to change the value based on window size.

File: assets/ui.css (hunk 7 above)
Comment (saschanaz) [reply assets/ui.css]: IMO it's okay as-is.

Comment (sidvishnoi) [reply assets/ui.css]: I prefer to keep `cursor:pointer` and remove the scaling animation.

**Comment by sidvishnoi** [comment]
Need to add `background: none; border: none` on button
image

File: src/core/ui.js [hunk 8]
```diff
@@ -47,7 +47,9 @@ function ariaDecorate(elem, ariaMap) {
 
 const respecUI = hyperHTML`<div id='respec-ui' class='removeOnSave' hidden></div>`;
 const menu = hyperHTML`<ul id=respec-menu role=menu aria-labelledby='respec-pill' hidden></ul>`;
+const closeButton = hyperHTML`<button class="close-button">❌</button>`;
```
Comment (sidvishnoi) [inline src/core/ui.js]:
```suggestion
const closeButton = hyperHTML`<button class="close-button" aria-label="Close">❌</button>`;
```

File: src/core/ui.js [hunk 9]
```diff
@@ -222,7 +224,7 @@ export const ui = {
     const id = `${currentOwner.id}-modal`;
     const headingId = `${id}-heading`;
     modal = hyperHTML`<div id='${id}' class='respec-modal removeOnSave' role='dialog'>
-      <h3 id="${headingId}">${title}</h3>
+      <h3 id="${headingId}">${closeButton}${title}</h3>
```
Comment (sidvishnoi) [inline src/core/ui.js]: I'm not sure of the accessibility implications of adding close button next to heading. It'll be announced as: "heading level 3 button close Search definitions"
Might be better to move this out of the heading and add a `aria-labelledby="${headingID}"` on `div[role='dialog']` above.

File: src/core/ui.js [hunk 10]
```diff
@@ -47,7 +47,9 @@ function ariaDecorate(elem, ariaMap) {
 
 const respecUI = hyperHTML`<div id='respec-ui' class='removeOnSave' hidden></div>`;
 const menu = hyperHTML`<ul id=respec-menu role=menu aria-labelledby='respec-pill' hidden></ul>`;
+const closeButton = hyperHTML`<button class="close-button" aria-label="Close">❌</button>`;
```
Comment (saschanaz) [inline src/core/ui.js]:
```suggestion
const closeButton = hyperHTML`<button class="close-button" title="Close">❌</button>`;
```

`title` also allows visual readers to see the button role by mouse hovering, while screen readers will still perceive it as `aria-label`.

Comment (sidvishnoi) [reply src/core/ui.js]: That's even better!

File: src/core/ui.js (hunk 9 above)
Comment (SuyashSalampuria) [reply src/core/ui.js]: Should the close button be at the same place?
Screenshot from 2020-01-09 13-48-19
Also, shall I write it something like this:
```
<div id="${headingId}">${closeButton} <h3>${title}</h3></div>
```

File: src/core/ui.js [hunk 11]
```diff
@@ -221,13 +223,14 @@ export const ui = {
     overlay = hyperHTML`<div id='respec-overlay' class='removeOnSave'></div>`;
     const id = `${currentOwner.id}-modal`;
     const headingId = `${id}-heading`;
-    modal = hyperHTML`<div id='${id}' class='respec-modal removeOnSave' role='dialog'>
+    modal = hyperHTML`<div id='${id}' class='respec-modal removeOnSave' role='dialog' aria-labelledby='${headingId}' >
```
Comment (saschanaz) [inline src/core/ui.js]:
```suggestion
    modal = hyperHTML`<div id='${id}' class='respec-modal removeOnSave' role='dialog' aria-labelledby='${headingId}' >
      ${closeButton}
```

I think this was the suggestion 👀

Comment (SuyashSalampuria) [reply src/core/ui.js]: I had tried that. But a white line was added above the heading line and that is offcourse not desirable
Screenshot from 2020-01-09 19-18-46

Comment (saschanaz) [reply src/core/ui.js]: Hmm, we might want to do `position: absolute` in that case. @sidvishnoi What do you think?

Comment (sidvishnoi) [reply src/core/ui.js]: The markup is what we want, so yes, position it with CSS.

**Comment by SuyashSalampuria** [comment]
> Why these changes?

<h3> always changes the line. So we need an inline display element

File: assets/ui.css [hunk 12]
```diff
@@ -231,7 +243,6 @@
   background: #fff;
   border: 5px solid #666;
   min-width: 20%;
-  width: 79%;
```
Comment (sidvishnoi) [inline assets/ui.css]: Why was this removed? I'm seeing modal like this now:
image

File: assets/ui.css [hunk 13]
```diff
@@ -245,6 +256,7 @@
 }
 
 .respec-modal h3 {
+  width: 100%;
```
Comment (sidvishnoi) [inline assets/ui.css]: This is causing an overflow-x. Any reason to add it?

File: assets/ui.css [hunk 14]
```diff
@@ -1,3 +1,15 @@
+.close-button {
+  position: absolute;
+  z-index: inherit;
+  padding-top: .2em;
+  font-weight: bold;
+  float: left;
```
Comment (sidvishnoi) [inline assets/ui.css]: We have `position: absolute` already
```suggestion
```

File: assets/ui.css [hunk 15]
```diff
@@ -1,3 +1,15 @@
+.close-button {
+  position: absolute;
+  z-index: inherit;
+  padding-top: .2em;
```
Comment (sidvishnoi) [inline assets/ui.css]:
```suggestion
  padding: .2em;
```

File: assets/ui.css (hunk 12 above)
Comment (SuyashSalampuria) [reply assets/ui.css]: There was a small gap between the heading and the boundary in some of the modals(search definitions). I wanted to remove that actually. Correcting this!

File: src/core/ui.js [hunk 16]
```diff
@@ -47,7 +47,9 @@ function ariaDecorate(elem, ariaMap) {
 
 const respecUI = hyperHTML`<div id='respec-ui' class='removeOnSave' hidden></div>`;
 const menu = hyperHTML`<ul id=respec-menu role=menu aria-labelledby='respec-pill' hidden></ul>`;
+const closeButton = hyperHTML`<button class="close-button" title="Close">❌</button>`;
 window.addEventListener("load", () => trapFocus(menu));
+closeButton.addEventListener("click", () => ui.closeModal());
```
Comment (sidvishnoi) [inline src/core/ui.js]: Can we do this inline?
``` js
hyperHTML`<button class="close-button" title="Close" onclick={() => ui.closeModal()}>❌</button>`
```

File: assets/ui.css [hunk 17]
```diff
@@ -1,3 +1,14 @@
+.close-button {
+  position: absolute;
```
Comment (sidvishnoi) [inline assets/ui.css]: Should we move this to right side? Feels more natural to me. WDYT @saschanaz ?

Comment (SuyashSalampuria) [reply assets/ui.css]: I was asked to add this at that place by @marcoscaceres

File: assets/ui.css [hunk 18]
```diff
@@ -245,6 +256,7 @@
 }
 
 .respec-modal h3 {
+  /* width: 100%; */
```
Comment (sidvishnoi) [inline assets/ui.css]:
```suggestion
```

File: assets/ui.css [hunk 19]
```diff
@@ -1,3 +1,14 @@
+.close-button {
```
Comment (sidvishnoi) [inline assets/ui.css]: Maybe just to be safe?
```suggestion
.respec-modal .close-button {
```

File: src/core/ui.js [hunk 20]
```diff
@@ -47,6 +47,8 @@ function ariaDecorate(elem, ariaMap) {
 
 const respecUI = hyperHTML`<div id='respec-ui' class='removeOnSave' hidden></div>`;
 const menu = hyperHTML`<ul id=respec-menu role=menu aria-labelledby='respec-pill' hidden></ul>`;
+const closeButton = hyperHTML`<button class="close-button" onclick=${() =>
```
Comment (sidvishnoi) [inline src/core/ui.js]: Nit: can we avoid creating another function?
```suggestion
const closeButton = hyperHTML`<button class="close-button" onclick=${ui.closeModal} title="Close">❌</button>`;
```

Comment (SuyashSalampuria) [reply src/core/ui.js]: We actually need a new function because in that case, ui will be used directly before its declaration

**Comment by SuyashSalampuria** [comment]
Can this be merged if it is ready because I need the close button for trapping focus inside modals

**Comment by sidvishnoi** [comment]
Missed that. Merged now :slightly_smiling_face:
